src/main/java/au/org/ala/biocache/web/CustomExceptionResolver.java

```diff
@@ -0,0 +1,42 @@
+/**************************************************************************
+ *  Copyright (C) 2013 Atlas of Living Australia
+ *  All Rights Reserved.
+ *
+ *  The contents of this file are subject to the Mozilla Public
+ *  License Version 1.1 (the "License"); you may not use this file
+ *  except in compliance with the License. You may obtain a copy of
+ *  the License at http://www.mozilla.org/MPL/
+ *
+ *  Software distributed under the License is distributed on an "AS
+ *  IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
+ *  implied. See the License for the specific language governing
+ *  rights and limitations under the License.
+ ***************************************************************************/
+package au.org.ala.biocache.web;
+
+import org.apache.solr.common.SolrException;
+import org.springframework.web.servlet.ModelAndView;
+import org.springframework.web.servlet.handler.SimpleMappingExceptionResolver;
+
+public class CustomExceptionResolver extends SimpleMappingExceptionResolver {
+    @Override
+    protected ModelAndView getModelAndView(String viewName, Exception ex) {
+        ModelAndView mv = new ModelAndView(viewName);
+
+        if (ex instanceof org.springframework.dao.DataAccessException ||
+            ex instanceof org.springframework.transaction.TransactionException) {
+            mv.addObject("error", ex);
+        } else {
+            // if user specified 'accept: application/json' in header, ContentNegotiatingViewResolver
+            // will try to convert the modal into json which could crash (exception can't  be mapped to a json value)
+            // so we construct a proper modal here
+            //
+            // if no 'application/json' specified, view resolver will resolve to jsp pages in which
+            // request.getAttribute("javax.servlet.error.exception") is used to get the exception and show details to user
+            // this function won't be affected by this custom class so jsp pages will work as before
+            mv.addObject("message", ex.getMessage());
+            mv.addObject("errorType", (ex instanceof SolrException) ? "Query syntax invalid" : "Server error");
```

Comment:
I'm not sure I'm down with this change (assuming I'm reading it right). It seems that any exception apart from `DataAccessException`, `TransactionException` & `SolrException` are getting a `Server error` type. What about other errors such as missing required params or 404, etc. They are not server errors but will appear as them. Also, what if the exception already has `errorType: `Some other message`? There might be duplicate keys in the JSON obj conversion and possibly throw another exception?

Reply:
Hi Nick, this is the setting before the change

```
<bean class="org.springframework.web.servlet.handler.SimpleMappingExceptionResolver">
    <property name="exceptionAttribute" value="error"/>
    <property name="defaultStatusCode" value="500"/>
    <property name="exceptionMappings">
        <props>
            <prop key="org.springframework.web.servlet.PageNotFound">error/pageNotFound</prop>
            <prop key="org.springframework.dao.DataAccessException">error/dataAccessFailure</prop>
            <prop key="org.springframework.transaction.TransactionException">error/dataAccessFailure</prop>
        </props>
    </property>
    <property name="statusCodes">
        <props>
            <prop key="error/pageNotFound">404</prop>
        </props>
    </property>
    <property name="defaultErrorView" value="error/general" />
</bean>
```
PageNotFound, DataAccessException and TransactionException go to their own pages, all other exceptions go to `error/general`, in which SolrException has a `errorType: Query syntax invalid`, all others have `errorType: Server error`

https://github.com/AtlasOfLivingAustralia/biocache-service/blob/develop/src/main/webapp/WEB-INF/jsp/error/general.jsp


---
the problem is `<prop key="org.springframework.web.servlet.PageNotFound">error/pageNotFound</prop>`
key part is supposed to be a kind of exception but actually org.springframework.web.servlet.PageNotFound is not an exception. I don't know where/how it can be thrown

Reply:
I think Spring generates the `org.springframework.web.servlet.PageNotFound` exception when a path is requested, that has no mapping to a controller method.

Reply:
I think this PR needs some integration tests for controller methods, so we can check different request types are handled by the expected response type and status.

Reply:
IntelliJ told me org.springframework.web.servlet.PageNotFound is not a type of exception, seems it's a String value.

Reply:
Theoretically we only changed the content of model in some cases ( instead of `mv.addObject("error", ex);` now we put our owns). This model won't be used in any jsp pages (they use ` request.getAttribute("javax.servlet.error.exception")` to get the exception and display)

They only will be used when `accept: application/json` is specified and so MappingJackson2JsonView is used


With this said, if anyone reaches the controller with `acception: application/json` and there's an exception. It's like this:

```
curl 'http://localhost:8078/occurrence/test/1' -H 'accept: application/json, text/plain, */*'
{
"message": "Direct self-reference leading to cycle (through reference chain: java.util.HashMap["error"]->org.springframework.transaction.TransactionTimedOutException["mostSpecificCause"])", "errorType": "Server error"
}
```

but if you do it in browser, you see

Internal Error
Data access failure has occurred. If this error persists you might like to let us know via email.

Reply:
So this issue can be summarized as: 

when there's an exception, `SimpleMappingExceptionResolver` directly puts the exception into the model. when `accept: application/json` is specified, this model will be converted to JSON and then could fail.

 we just need to make sure the model can be converted to JSON successfully in this case.